File: data/skills/scrnaseq-scanpy-core-analysis/scripts/find_markers.py
"""
============================================================================
MARKER GENE IDENTIFICATION
============================================================================

This script identifies marker genes for each cluster.

Functions:
  - find_all_cluster_markers(): Find markers for all clusters
  - find_markers_for_cluster(): Find markers for a specific cluster
  - export_marker_tables(): Export marker gene tables
  - plot_top_markers_heatmap(): Heatmap of top marker genes
  - plot_markers_dotplot(): Dot plot of top marker genes

Usage:
  from find_markers import find_all_cluster_markers, plot_top_markers_heatmap
  markers = find_all_cluster_markers(adata, cluster_key='leiden_0.8')
  plot_top_markers_heatmap(adata, markers, output_dir='results/markers')
"""

from pathlib import Path
from typing import Dict, List, Optional, Union

import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns


def _save_plot(fig: plt.Figure, base_path: Union[str, Path], dpi: int = 300) -> None:
    """
    Save plot in both PNG and SVG formats with graceful fallback.

    Parameters
    ----------
    fig : matplotlib.figure.Figure
        Figure object to save
    base_path : str or Path
        Base path for output files (without extension)
    dpi : int, optional
        Resolution for PNG (default: 300)

    Returns
    -------
    None
        Saves files to disk
    """
    base_path = Path(base_path)

    # Always save PNG
    png_path = base_path.with_suffix('.png')
    try:
        fig.savefig(png_path, dpi=dpi, bbox_inches='tight', format='png')
        print(f"  Saved: {png_path}")
    except Exception as e:
        print(f"  Warning: PNG export failed: {e}")

    # Always try SVG
    svg_path = base_path.with_suffix('.svg')
    try:
        fig.savefig(svg_path, bbox_inches='tight', format='svg')
        print(f"  Saved: {svg_path}")
    except Exception as e:
        print(f"  (SVG export failed, PNG available)")


def find_all_cluster_markers(
    adata: 'AnnData',
    cluster_key: str = 'leiden_0.8',
    method: str = 'wilcoxon',
    n_genes: Optional[int] = None,
    min_in_group_fraction: float = 0.25,
    min_fold_change: float = 1.0,
    max_out_group_fraction: float = 0.5,
    use_raw: bool = False,
    layer: Optional[str] = None
) -> pd.DataFrame:
    """
    Find marker genes for all clusters.

    Parameters
    ----------
    adata : AnnData
        AnnData object
    cluster_key : str, optional
        Cluster column in adata.obs (default: 'leiden_0.8')
    method : str, optional
        Statistical test: 'wilcoxon', 't-test', 'logreg' (default: 'wilcoxon')
    n_genes : int, optional
        Number of genes to return per cluster (default: None, returns all)
    min_in_group_fraction : float, optional
        Minimum fraction of cells in group expressing gene (default: 0.25)
    min_fold_change : float, optional
        Minimum fold change (in natural log) (default: 1.0)
    max_out_group_fraction : float, optional
        Maximum fraction of cells outside group expressing gene (default: 0.5)
    use_raw : bool, optional
        Use raw counts (default: False)
    layer : str, optional
        Layer to use (default: None)

    Returns
    -------
    DataFrame
        Marker genes for all clusters
    """
    import scanpy as sc

    if cluster_key not in adata.obs.columns:
        raise ValueError(f"{cluster_key} not found in adata.obs")

    print(f"Finding marker genes for all clusters in '{cluster_key}'...")
    print(f"  Method: {method}")
    print(f"  Min in-group fraction: {min_in_group_fraction}")
    print(f"  Min fold change: {min_fold_change}")

    # Run marker gene identification
    sc.tl.rank_genes_groups(
        adata,
        groupby=cluster_key,
        method=method,
        use_raw=use_raw,
        layer=layer,
        n_genes=n_genes
    )

    # Filter markers
    sc.tl.filter_rank_genes_groups(
        adata,
        min_in_group_fraction=min_in_group_fraction,
        min_fold_change=min_fold_change,
        max_out_group_fraction=max_out_group_fraction
    )

    # Convert to dataframe
    result = sc.get.rank_genes_groups_df(adata, group=None)

    n_clusters = adata.obs[cluster_key].nunique()
    print(f"  Found markers for {n_clusters} clusters")
    print(f"  Total markers: {len(result)}")

    return result


def find_markers_for_cluster(
    adata: 'AnnData',
    cluster: str,
    cluster_key: str = 'leiden_0.8',
    method: str = 'wilcoxon',
    n_genes: int = 100
) -> pd.DataFrame:
    """
    Find marker genes for a specific cluster.

    Parameters
    ----------
    adata : AnnData
        AnnData object
    cluster : str
        Cluster ID
    cluster_key : str, optional
        Cluster column in adata.obs (default: 'leiden_0.8')
    method : str, optional
        Statistical test (default: 'wilcoxon')
    n_genes : int, optional
        Number of genes to return (default: 100)

    Returns
    -------
    DataFrame
        Marker genes for the cluster
    """
    import scanpy as sc

    if cluster_key not in adata.obs.columns:
        raise ValueError(f"{cluster_key} not found in adata.obs")

    print(f"Finding marker genes for cluster {cluster}...")

    # Run marker gene identification
    sc.tl.rank_genes_groups(
        adata,
        groupby=cluster_key,
        method=method,
        n_genes=n_genes
    )

    # Get results for specific cluster
    result = sc.get.rank_genes_groups_df(adata, group=cluster)

    print(f"  Found {len(result)} markers")

    return result


def export_marker_tables(
    markers: pd.DataFrame,
    output_dir: Union[str, Path] = ".",
    top_n: Optional[int] = None
) -> None:
    """
    Export marker gene tables to CSV files.

    Parameters
    ----------
    markers : DataFrame
        Marker gene dataframe from find_all_cluster_markers
    output_dir : str or Path, optional
        Output directory (default: ".")
    top_n : int, optional
        Export only top N markers per cluster (default: None, exports all)

    Returns
    -------
    None
        Saves CSV files to output_dir
    """
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    print(f"Exporting marker tables to {output_dir}...")

    # Export all markers
    if top_n is not None:
        markers_filtered = markers.groupby('group').head(top_n)
        output_file = output_dir / f"cluster_markers_top{top_n}.csv"
    else:
        markers_filtered = markers
        output_file = output_dir / "cluster_markers_all.csv"

    markers_filtered.to_csv(output_file, index=False)
    print(f"  Saved: {output_file}")

    # Export separate file for each cluster
    for cluster in markers['group'].unique():
        cluster_markers = markers[markers['group'] == cluster]
        if top_n is not None:
            cluster_markers = cluster_markers.head(top_n)

        output_file = output_dir / f"markers_cluster_{cluster}.csv"
        cluster_markers.to_csv(output_file, index=False)

    print(f"  Saved individual files for {markers['group'].nunique()} clusters")


def plot_top_markers_heatmap(
    adata: 'AnnData',
    markers: pd.DataFrame,
    n_top: int = 10,
    cluster_key: str = 'leiden_0.8',
    output_dir: Union[str, Path] = ".",
    figsize: tuple = (12, 8),
    use_raw: bool = False
) -> None:
    """
    Create heatmap of top marker genes using seaborn clustermap.

    Parameters
    ----------
    adata : AnnData
        AnnData object
    markers : DataFrame
        Marker gene dataframe
    n_top : int, optional
        Number of top markers per cluster (default: 10)
    cluster_key : str, optional
        Cluster column in adata.obs (default: 'leiden_0.8')
    output_dir : str or Path, optional
        Output directory (default: ".")
    figsize : tuple, optional
        Figure size (default: (12, 8))
    use_raw : bool, optional
        Use raw counts (default: False)

    Returns
    -------
    None
        Saves heatmap to output_dir
    """
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    print(f"Creating marker gene heatmap (top {n_top} per cluster)...")

    # Get top markers per cluster
    top_markers = markers.groupby('group').head(n_top)
    marker_genes = top_markers['names'].unique()

    print(f"  Plotting {len(marker_genes)} marker genes")

    # Get expression data
    if use_raw and adata.raw is not None:
        expr_data = adata.raw[:, marker_genes].X
    else:
        expr_data = adata[:, marker_genes].X

    # Convert to dense if sparse
    if hasattr(expr_data, 'toarray'):
        expr_data = expr_data.toarray()

    # Create dataframe
    expr_df = pd.DataFrame(
        expr_data,
        index=adata.obs_names,
        columns=marker_genes
    )

    # Add cluster annotations
    expr_df['cluster'] = adata.obs[cluster_key].values

    # Sort by cluster
    expr_df = expr_df.sort_values('cluster')

    # Prepare for heatmap
    row_colors = expr_df['cluster'].astype('category').cat.codes
    expr_df = expr_df.drop('cluster', axis=1)

    # Create heatmap
    g = sns.clustermap(
        expr_df.T,
        col_cluster=False,
        row_cluster=True,
        cmap='RdBu_r',
        center=0,
        figsize=figsize,
        cbar_kws={'label': 'Expression'},
        yticklabels=True,
        xticklabels=False
    )

    output_file = output_dir / "markers_heatmap"
    fig = plt.gcf()
    _save_plot(fig, output_file, dpi=300)
    plt.close()


def plot_markers_dotplot(
    adata: 'AnnData',
    markers: pd.DataFrame,
    n_top: int = 5,
    cluster_key: str = 'leiden_0.8',
    output_dir: Union[str, Path] = ".",
    figsize: tuple = (12, 6)
) -> None:
    """
    Create dot plot of top marker genes.

    Parameters
    ----------
    adata : AnnData
        AnnData object
    markers : DataFrame
        Marker gene dataframe
    n_top : int, optional
        Number of top markers per cluster (default: 5)
    cluster_key : str, optional
        Cluster column in adata.obs (default: 'leiden_0.8')
    output_dir : str or Path, optional
        Output directory (default: ".")
    figsize : tuple, optional
        Figure size (default: (12, 6))

    Returns
    -------
    None
        Saves dot plot to output_dir
    """
    import scanpy as sc

    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    print(f"Creating marker gene dot plot (top {n_top} per cluster)...")

    # Get top markers per cluster
    top_markers = markers.groupby('group').head(n_top)
    marker_genes = top_markers['names'].unique()

    print(f"  Plotting {len(marker_genes)} marker genes")

    sc.pl.dotplot(
        adata,
        var_names=marker_genes,
        groupby=cluster_key,
        dendrogram=True,
        show=False
    )

    output_file = output_dir / "markers_dotplot"
    fig = plt.gcf()
    _save_plot(fig, output_file, dpi=300)
    plt.close()

File: data/skills/scrnaseq-seurat-core-analysis/scripts/find_markers.R
# ============================================================================
# MARKER GENE IDENTIFICATION
# ============================================================================
#
# Find genes that define each cluster through differential expression.
#
# Functions:
#   - find_all_cluster_markers(): Find markers for all clusters
#   - find_markers_for_cluster(): Find markers for specific cluster
#   - export_marker_tables(): Save marker tables to CSV
#   - plot_top_markers_heatmap(): Heatmap of top markers
#   - plot_markers_dotplot(): Dot plot of top markers
#   - plot_markers_violin(): Violin plots of top markers
#
# Usage:
#   source("scripts/find_markers.R")
#   all_markers <- find_all_cluster_markers(seurat_obj)

library(ggplot2)
library(ggprism)
library(patchwork)
library(dplyr)

#' Find markers for all clusters
#'
#' @param seurat_obj Seurat object with clusters
#' @param resolution Resolution to use (NULL for active idents)
#' @param only_pos Only return positive markers (default: TRUE)
#' @param min_pct Minimum fraction of cells expressing gene (default: 0.25)
#' @param logfc_threshold Minimum log2 fold change (default: 0.25)
#' @param test_use DE test to use (default: "wilcox")
#' @param verbose Print progress (default: TRUE)
#' @return Data frame with marker genes
#' @export
find_all_cluster_markers <- function(seurat_obj,
                                    resolution = NULL,
                                    only_pos = TRUE,
                                    min_pct = 0.25,
                                    logfc_threshold = 0.25,
                                    test_use = "wilcox",
                                    verbose = TRUE) {

  message("Finding markers for all clusters")

  # Set identity if resolution specified
  if (!is.null(resolution)) {
    col_name <- paste0("RNA_snn_res.", resolution)
    if (col_name %in% colnames(seurat_obj@meta.data)) {
      Idents(seurat_obj) <- col_name
      message("  Using resolution: ", resolution)
    } else {
      stop("Resolution column not found: ", col_name)
    }
  }

  n_clusters <- length(unique(Idents(seurat_obj)))
  message("  Number of clusters: ", n_clusters)
  message("  Test: ", test_use)
  message("  Min pct: ", min_pct)
  message("  Log2FC threshold: ", logfc_threshold)

  # Find all markers
  all_markers <- FindAllMarkers(
    seurat_obj,
    only.pos = only_pos,
    min.pct = min_pct,
    logfc.threshold = logfc_threshold,
    test.use = test_use,
    verbose = verbose
  )

  # Add cluster info
  all_markers <- all_markers %>%
    arrange(cluster, desc(avg_log2FC))

  message("Marker identification complete")
  message("  Total markers found: ", nrow(all_markers))
  message("  Markers per cluster:")
  print(table(all_markers$cluster))

  return(all_markers)
}

#' Find markers for specific cluster
#'
#' @param seurat_obj Seurat object with clusters
#' @param cluster_id Cluster ID to find markers for
#' @param resolution Resolution to use (NULL for active idents)
#' @param only_pos Only return positive markers (default: TRUE)
#' @param min_pct Minimum fraction of cells expressing gene (default: 0.25)
#' @param logfc_threshold Minimum log2 fold change (default: 0.25)
#' @return Data frame with marker genes
#' @export
find_markers_for_cluster <- function(seurat_obj,
                                    cluster_id,
                                    resolution = NULL,
                                    only_pos = TRUE,
                                    min_pct = 0.25,
                                    logfc_threshold = 0.25) {

  message("Finding markers for cluster ", cluster_id)

  # Set identity if resolution specified
  if (!is.null(resolution)) {
    col_name <- paste0("RNA_snn_res.", resolution)
    Idents(seurat_obj) <- col_name
  }

  # Find markers
  markers <- FindMarkers(
    seurat_obj,
    ident.1 = cluster_id,
    only.pos = only_pos,
    min.pct = min_pct,
    logfc.threshold = logfc_threshold
  )

  # Add gene names
  markers$gene <- rownames(markers)
  markers <- markers %>%
    arrange(desc(avg_log2FC))

  message("  Markers found: ", nrow(markers))

  return(markers)
}

#' Export marker tables to CSV
#'
#' @param all_markers Data frame from find_all_cluster_markers()
#' @param output_dir Output directory
#' @param top_n Number of top markers per cluster (default: 50)
#' @export
export_marker_tables <- function(all_markers, output_dir, top_n = 50) {

  message("Exporting marker tables")

  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

  # Export all markers
  all_file <- file.path(output_dir, "cluster_markers_all.csv")
  write.csv(all_markers, all_file, row.names = FALSE)
  message("  Saved: ", all_file)

  # Export top N per cluster
  top_markers <- all_markers %>%
    group_by(cluster) %>%
    slice_head(n = top_n)

  top_file <- file.path(output_dir, paste0("cluster_markers_top", top_n, ".csv"))
  write.csv(top_markers, top_file, row.names = FALSE)
  message("  Saved: ", top_file)

  # Export separate file per cluster
  for (clust in unique(all_markers$cluster)) {
    clust_markers <- all_markers %>%
      filter(cluster == clust)

    clust_file <- file.path(output_dir, paste0("cluster_", clust, "_markers.csv"))
    write.csv(clust_markers, clust_file, row.names = FALSE)
  }

  message("  Individual cluster files saved")
}

#' Plot heatmap of top markers
#'
#' @param seurat_obj Seurat object
#' @param all_markers Data frame from find_all_cluster_markers()
#' @param n_top Number of top markers per cluster (default: 10)
#' @param output_dir Output directory for plots
#' @return Heatmap object
#' @export
plot_top_markers_heatmap <- function(seurat_obj,
                                    all_markers,
                                    n_top = 10,
                                    output_dir = NULL) {

  message("Creating marker heatmap")

  # Get top markers per cluster
  top_markers <- all_markers %>%
    group_by(cluster) %>%
    slice_head(n = n_top) %>%
    pull(gene)

  message("  Plotting ", length(top_markers), " markers")

  # Create heatmap
  p <- DoHeatmap(
    seurat_obj,
    features = top_markers,
    size = 3
  ) +
    theme_prism() +
    theme(axis.text.y = element_text(size = 6))

  # Save plot if output directory specified
  if (!is.null(output_dir)) {
    dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

    svg_file <- file.path(output_dir, "markers_heatmap.svg")
    ggsave(svg_file, plot = p, width = 12, height = 14, dpi = 300)
    message("  Saved: ", svg_file)

    png_file <- file.path(output_dir, "markers_heatmap.png")
    ggsave(png_file, plot = p, width = 12, height = 14, dpi = 300)
    message("  Saved: ", png_file)
  }

  return(p)
}

#' Plot dot plot of top markers
#'
#' @param seurat_obj Seurat object
#' @param all_markers Data frame from find_all_cluster_markers()
#' @param n_top Number of top markers per cluster (default: 5)
#' @param output_dir Output directory for plots
#' @return ggplot object
#' @export
plot_markers_dotplot <- function(seurat_obj,
                                all_markers,
                                n_top = 5,
                                output_dir = NULL) {

  message("Creating marker dot plot")

  # Get top markers per cluster
  top_markers <- all_markers %>%
    group_by(cluster) %>%
    slice_head(n = n_top) %>%
    pull(gene) %>%
    unique()

  message("  Plotting ", length(top_markers), " markers")

  # Create dot plot
  p <- DotPlot(
    seurat_obj,
    features = top_markers
  ) +
    theme_prism() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
    labs(title = "Top Marker Genes per Cluster")

  # Save plot if output directory specified
  if (!is.null(output_dir)) {
    dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

    svg_file <- file.path(output_dir, "markers_dotplot.svg")
    ggsave(svg_file, plot = p, width = 12, height = 6, dpi = 300)
    message("  Saved: ", svg_file)

    png_file <- file.path(output_dir, "markers_dotplot.png")
    ggsave(png_file, plot = p, width = 12, height = 6, dpi = 300)
    message("  Saved: ", png_file)
  }

  return(p)
}

#' Plot violin plots of top markers
#'
#' @param seurat_obj Seurat object
#' @param all_markers Data frame from find_all_cluster_markers()
#' @param n_top Number of top markers per cluster (default: 3)
#' @param output_dir Output directory for plots
#' @return Combined ggplot object
#' @export
plot_markers_violin <- function(seurat_obj,
                               all_markers,
                               n_top = 3,
                               output_dir = NULL) {

  message("Creating marker violin plots")

  # Get top markers per cluster
  top_markers <- all_markers %>%
    group_by(cluster) %>%
    slice_head(n = n_top) %>%
    pull(gene) %>%
    unique()

  # Limit to reasonable number
  if (length(top_markers) > 20) {
    message("  Too many markers, limiting to 20")
    top_markers <- head(top_markers, 20)
  }

  message("  Plotting ", length(top_markers), " markers")

  # Create violin plot
  p <- VlnPlot(
    seurat_obj,
    features = top_markers,
    ncol = 4,
    pt.size = 0
  ) &
    theme_prism() &
    theme(
      legend.position = "none",
      axis.title.x = element_blank(),
      axis.text.x = element_text(angle = 45, hjust = 1)
    )

  # Save plot if output directory specified
  if (!is.null(output_dir)) {
    dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

    # Calculate height based on number of features
    nrow <- ceiling(length(top_markers) / 4)
    height <- nrow * 3

    svg_file <- file.path(output_dir, "markers_violin.svg")
    ggsave(svg_file, plot = p, width = 16, height = height, dpi = 300)
    message("  Saved: ", svg_file)

    png_file <- file.path(output_dir, "markers_violin.png")
    ggsave(png_file, plot = p, width = 16, height = height, dpi = 300)
    message("  Saved: ", png_file)
  }

  return(p)
}
